vault/auth.go

```diff
@@ -81,9 +81,13 @@ func (c *Core) enableCredentialInternal(ctx context.Context, entry *MountEntry,
 
 	c.mountsLock.Lock()
 	c.authLock.Lock()
+	locked := true
 	unlock := func() {
-		c.authLock.Unlock()
-		c.mountsLock.Unlock()
+		if locked {
+			c.authLock.Unlock()
+			c.mountsLock.Unlock()
+			locked = false
+		}
```

Comment:
I think this is more readable. Now we are relying on mutating `locked` rather than `unlock` it would be possible to do this without the named variable. I'm in two minds about which is clearer though so happy to leave it like this if you think this is more readable:

```go
locked := true
defer func(){
  if locked {
    // ...
  }
}()
```

I don't mind the name and in some ways like that it's expressing intent... but the fact it's a variable makes me think twice about the execution behaviour in the defer below (i.e. my brain might spend cycles convincing myself there is no subtle bug like the one we are fixing 😄 )

What do you think?

Reply:
I think the main reason for the named variable function is that `unlock` is called manually in line 230 after which `disableCredentialInternal` is executed. I have not looked into it in depth, but it seems like the order matters?